colexec: call DrainMeta and Next from single goroutine

asubiotto opened the pull request:
Downstream operators would previously call DrainMeta on an operator whose Next method was called from a different goroutine, which was safe because the interface contract stipulates that these implementations handle concurrent calls to Next and DrainMeta. However, this led to complicated DrainMeta implementations (see `colrpc/inbox.go` and sometimes incorrect behavior (#48785).

This PR implements the MetadataSource interface in the two components that would create internal goroutines: HashRouter and ParallelUnorderedSynchronizer. Metadata is now safely passed between internal goroutines so that DrainMeta can be called on these components instead of fetching metadata directly from an operator running in another goroutine. This effectively pushes the synchronization overhead to these components, allowing us to simplify the DrainMeta contract and fixing bugs caused by the previous model.

Release note: None

Closes #48785
Fixes #49642 
Fixes #49741

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50009)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Very nice work and great tests!

> TODO: Does this need a release note? Did the short-circuiting change make it to a release?

I don't think it needs a release note, and the short-circuiting didn't make it into any release (even alpha).

The commit message from the second commit seems to imply that there is some follow-up work left, is there?

> Reviewed 6 of 6 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-:-M9POVkX3jVxTD5KM2vH:bu2p4lz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/ordered_synchronizer_tmpl.go, line 239 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PiE0p9xnRb7FImGgo:-M9PiE0p9xnRb7FImGgp:bb6yhwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/ordered_synchronizer_tmpl.go#L239)):*
> ```Go
> 	var bufferedMeta []execinfrapb.ProducerMetadata
> 	for _, input := range o.inputs {
> 		if meta := input.MetadataSources.DrainMeta(ctx); len(meta) != 0 {
> ```

ditto for the length check.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PiVX-B27X-8BwHuwN:-M9PiVX-B27X-8BwHuwO:bgnn455) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L40)):*
> ```Go
> var _ execinfra.OpNode = &ParallelUnorderedSynchronizer{}
> 
> type parallelUnorderedSynchronizerState int
> ```

ditto for possible change from `int` to `int32` to be reused below for `state` variable. Or maybe the current is actually a good idea since it kind of forces the reader of the code to use "get state" and "set state" methods that provide atomicity. Up to you.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 241 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9Pjei2D_Ayf_QpQe69:-M9Pjei2D_Ayf_QpQe6A:b3xrdgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L241)):*
> ```Go
> 					if input.MetadataSources != nil {
> 						msg.meta = input.MetadataSources[i].DrainMeta(ctx)
> 						sentMeta = true
> ```

This is slightly confusing. I think `sentMeta` should be updated below, in `case s.batchCh <- msg` when `msg.meta != nil` since that's where we actually send the metadata.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 276 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9Pk1Us-q5aN-olKFaR:-M9Pk1Us-q5aN-olKFaS:b-p56mtg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L276)):*
> ```Go
> 	for {
> 		state := s.getState()
> 		fmt.Printf("state is %d\n", state)
> ```

Leftover artifact.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 317 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PkCJB7Bw-vSvKlR9y:-M9PkCJB7Bw-vSvKlR9z:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L317)):*
> ```Go
> 			s.lastReadInputIdx = msg.inputIdx
> 			if msg.meta != nil {
> 				fmt.Printf("appending meta from %d and continuing\n", msg.inputIdx)
> ```

ditto
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 348 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PkcY48jXMQqui8a2Y:-M9PkcY48jXMQqui8a2Z:b-34xpmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L348)):*
> ```Go
> 		s.init(ctx)
> 	} else {
> 		// Some data must have been read in Next since the inputs were initialized.
> ```

Is this guaranteed that some data has been read? I think it is plausible that `init` has been called (which spawns all goroutines), but `ParallelUnorderedSynchronizer.DrainMeta` is called before any of those goroutines were able to actually read anything. In that case `lastReadInputIdx` will be left unchanged as zero. To be clear, I think the code itself is correct because we could be notifying the zeroth input in such scenario, just the comment might need an adjustment. Or maybe I'm missing something.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer_test.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PmIIO1pXXQUnfqG1K:-M9PmIIO1pXXQUnfqG1L:b-aan68z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer_test.go#L70)):*
> ```Go
> 
> 	ctx, cancelFn := context.WithCancel(context.Background())
> 	type parallelUnorderedSynchronizerTerminationScenario int
> ```

nit: this name seems rather long for a type that is defined within a test function, I'd probably just use "terminationScenario" or "scenario".
___
*[pkg/sql/colexec/parallel_unordered_synchronizer_test.go, line 107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9Pn-t23v8czZBpLWol:-M9Pn-t23v8czZBpLWom:bpe183k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer_test.go#L107)):*
> ```Go
> 			expectZeroBatch := false
> 			if terminationScenario == parallelUnorderedSynchronizerPrematureDrainMeta && batchesReturned < expectedBatchesReturned {
> 				// Call DrainMeta before the input is finished. Intentionally call Next
> ```

super nit: I'd explicitly say something like "Intentionally allow for Next to be called below even though ...".
___
*[pkg/sql/colexec/parallel_unordered_synchronizer_test.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PnAWD6q0M5GN7nt_L:-M9PnAWD6q0M5GN7nt_M:bh5vkhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer_test.go#L124)):*
> ```Go
> 			}
> 			if b.Length() == 0 {
> 				// Successful run, check that all inputs have returned metadata.
> ```

nit: this comment probably should be moved one line below.
___
*[pkg/sql/colexec/routers.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PQ0aV7eEfJngyzsvF:-M9PQ0aV7eEfJngyzsvG:b-ngcg7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L241)):*
> ```Go
> // nextErrorLocked is a helper method that handles an error encountered in Next.
> func (o *routerOutputOp) nextErrorLocked(ctx context.Context, err error) {
> 	fmt.Println("output calling next error", err)
> ```

This should probably logged and not printed.
___
*[pkg/sql/colexec/routers.go, line 275 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PQdkN4qJrngtHWRrU:-M9PQdkN4qJrngtHWRrV:bzhnaqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L275)):*
> ```Go
> 		var err error
> 		b, err = o.mu.data.dequeue(ctx)
> 		if o.testingKnobs.nextTestInducedErrorCb != nil && err == nil {
> ```

nit: in a similar condition below we check `err` first, I think we should do the same here.
___
*[pkg/sql/colexec/routers.go, line 455 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PRBK01iJX3nIWhN7W:-M9PRBK01iJX3nIWhN7X:b-g1x5un) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L455)):*
> ```Go
> }
> 
> type hashRouterDrainState int
> ```

nit: maybe `s/hashRouterDrainState/hashRouterState/g`? Then we would have `hashRouterStateRunning`, `hashRouterStateDrainRequested`, `hashRouterStateDrainCompleted` (the last one could be `hashRouterStateDone` or "finished).
___
*[pkg/sql/colexec/routers.go, line 502 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PRY5bBbKJNCGcaGqS:-M9PRY5bBbKJNCGcaGqT:b5hwany) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L502)):*
> ```Go
> 		// drainState is the state the hashRouter is in. The Run goroutine should
> 		// only ever read these states, never set them.
> 		drainState        int32
> ```

Probably `s/int32/hashRouterDrainState/`. Oh, this will probably require changing `type hashRouterDrainState` to be `int32` too. Up to you if we want to do this.
___
*[pkg/sql/colexec/routers.go, line 744 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PStpL46EikBO3F5H9:-M9PStpL46EikBO3F5HA:b-smnpmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L744)):*
> ```Go
> 	}
> 	r.setDrainState(hashRouterDrainStateRunning)
> 	r.waitForMetadata = make(chan []execinfrapb.ProducerMetadata)
> ```

Do we want to specify 1 as the size of the channel?
___
*[pkg/sql/colexec/routers.go, line 771 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PTHLY7zn90oiwK9Bn:-M9PTHLY7zn90oiwK9Bo:b8xrcpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L771)):*
> ```Go
> 		return nil
> 	}
> 	// All outputs have drained, return any buffered metadata to the last output
> ```

nit: probably `s/have drained/have been drained/`.
___
*[pkg/sql/colexec/routers_test.go, line 704 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PTivsB23OK6WLEDtj:-M9PTivsB23OK6WLEDtk:b-wlkcyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L704)):*
> ```Go
> 		for range outputs {
> 			if routerMeta := r.drainMeta(); routerMeta != nil {
> 				if metadata != nil {
> ```

Should we also check that the last output in particular returned the metadata?
___
*[pkg/sql/colexec/routers_test.go, line 759 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PTwsn0MdB3r6gu0pp:-M9PTwsn0MdB3r6gu0pq:b4o3121) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L759)):*
> ```Go
> 			}
> 
> 			routerMeta := make(chan []execinfrapb.ProducerMetadata)
> ```

Is this still used?
___
*[pkg/sql/colexec/routers_test.go, line 984 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PdlFiDGec7XgVs7XY:-M9PdlFiDGec7XgVs7XZ:b-2ny66l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L984)):*
> ```Go
> 							addBatchRng, _ := randutil.NewPseudoRand()
> 							// Sleep between 0 and ~5 milliseconds.
> 							fmt.Println("in add batch error")
> ```

Seems like a leftover artifact.
___
*[pkg/sql/colexec/routers_test.go, line 1060 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PeWrm9dorv8aKLApO:-M9PeWrm9dorv8aKLApP:b-5u6s9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L1060)):*
> ```Go
> 				wg.Add(1)
> 				go func() {
> 					if terminationScenario == hashRouterContextCanceled || terminationScenario == hashRouterChaos {
> ```

Probably we could use `isTerminationScenario` with 1.0 as probability.
___
*[pkg/sql/colexec/routers_test.go, line 1237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PfcqyBiZrowat8QCs:-M9PfcqyBiZrowat8QCt:b-6ly7jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L1237)):*
> ```Go
> 								actualDistribution[j] += oBatch.Length()
> 								if oBatch.Length() == 0 {
> 									_ = outputs[j].DrainMeta(ctx)
> ```

nit: we probably can omit `_ = ` part, right?
___
*[pkg/sql/colexec/serial_unordered_synchronizer.go, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PnKDL0U8E09IiH3ll:-M9PnKDL0U8E09IiH3lm:bb6yhwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/serial_unordered_synchronizer.go#L82)):*
> ```Go
> 	var bufferedMeta []execinfrapb.ProducerMetadata
> 	for _, input := range s.inputs {
> 		if meta := input.MetadataSources.DrainMeta(ctx); len(meta) != 0 {
> ```

ditto for the length check.
___
*[pkg/sql/execinfrapb/data.go, line 395 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9Phaee7zZrm8F28L0j:-M9Phaee7zZrm8F28L0k:b-te039z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/execinfrapb/data.go#L395)):*
> ```Go
> 	var result []ProducerMetadata
> 	for _, src := range s {
> 		if meta := src.DrainMeta(ctx); len(meta) != 0 {
> ```

nit: we don't need to check the length, can just append unconditionally (I've just confirmed it here https://play.golang.org/p/d41I1wL7kIg)


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
TFTR!

The followup items are: 1) Make sure that the contract is the same for `IdempotentClose` and 2) Simplify `DrainMeta` in a couple of components that were previously protecting from concurrent calls (e.g. inbox). I think I'm going to tackle those in a separate PR though to keep the amount of code manageable.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-:-M9Xn1DxDUUjuhskmFdW:brg0puu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/ordered_synchronizer_tmpl.go, line 239 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PiE0p9xnRb7FImGgo:-M9Y8Qj_2nH7px0faYNF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/ordered_synchronizer_tmpl.go#L239)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto for the length check.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PiVX-B27X-8BwHuwN:-M9Y8DCkENksunGOA_1G:bcxrvh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L40)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto for possible change from `int` to `int32` to be reused below for `state` variable. Or maybe the current is actually a good idea since it kind of forces the reader of the code to use "get state" and "set state" methods that provide atomicity. Up to you.
</blockquote></details>

Yeah, we need to have atomicity and changing it to an `int` doesn't change the number of casts we need to do.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 241 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9Pjei2D_Ayf_QpQe69:-M9Y89X--hjbsmiZbHsh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L241)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This is slightly confusing. I think `sentMeta` should be updated below, in `case s.batchCh <- msg` when `msg.meta != nil` since that's where we actually send the metadata.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 276 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9Pk1Us-q5aN-olKFaR:-M9Y7qca8RVhyVa9r9CY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L276)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Leftover artifact.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 317 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PkCJB7Bw-vSvKlR9y:-M9Y7psfBjjUIQXDryvv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L317)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 348 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PkcY48jXMQqui8a2Y:-M9Y6F84-qfW9_7SwMrh:b-x07o7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer.go#L348)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Is this guaranteed that some data has been read? I think it is plausible that `init` has been called (which spawns all goroutines), but `ParallelUnorderedSynchronizer.DrainMeta` is called before any of those goroutines were able to actually read anything. In that case `lastReadInputIdx` will be left unchanged as zero. To be clear, I think the code itself is correct because we could be notifying the zeroth input in such scenario, just the comment might need an adjustment. Or maybe I'm missing something.
</blockquote></details>

You're right, updated the comment.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer_test.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PmIIO1pXXQUnfqG1K:-M9Y59M1EMT5CgI_Np5_:b-ie9lu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer_test.go#L70)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: this name seems rather long for a type that is defined within a test function, I'd probably just use "terminationScenario" or "scenario".
</blockquote></details>

Renamed to `syncrhonizerTerminationScenario`
___
*[pkg/sql/colexec/parallel_unordered_synchronizer_test.go, line 107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9Pn-t23v8czZBpLWol:-M9Y4xh-Bu2Vm5ejUHvC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/parallel_unordered_synchronizer_test.go#L107)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: I'd explicitly say something like "Intentionally allow for Next to be called below even though ...".
</blockquote></details>

Done.
___
*[pkg/sql/colexec/routers.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PQ0aV7eEfJngyzsvF:-M9XndnZ7RzCKN-9Y5ZQ:b-6vmxm1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L241)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This should probably logged and not printed.
</blockquote></details>

This was a leftover artifact. I don't think it's necessary to log the error since it'll be returned in Next.
___
*[pkg/sql/colexec/routers.go, line 455 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PRBK01iJX3nIWhN7W:-M9Xnzx50c1yH2sbNmAx:bi8uqop) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L455)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: maybe `s/hashRouterDrainState/hashRouterState/g`? Then we would have `hashRouterStateRunning`, `hashRouterStateDrainRequested`, `hashRouterStateDrainCompleted` (the last one could be `hashRouterStateDone` or "finished).
</blockquote></details>

I started off with `hashRouterState` but I found it confusing because these states are explicitly about draining. For example, a hash router can have drained and exited the `Run` method but still be in `hashRouterStateRunning` until a somebody calls `drainMeta`. I'm happy to change it back, but wondering what you think.
___
*[pkg/sql/colexec/routers.go, line 502 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PRY5bBbKJNCGcaGqS:-M9XoPXG-4QWJoP2QLPj:b-irnlsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L502)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Probably `s/int32/hashRouterDrainState/`. Oh, this will probably require changing `type hashRouterDrainState` to be `int32` too. Up to you if we want to do this.
</blockquote></details>

It needs to be an `int32` to be able to use `atomic` on it. This also makes it explicit that it shouldn't be accessed directly (since doing so would require a cast).
___
*[pkg/sql/colexec/routers.go, line 744 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PStpL46EikBO3F5H9:-M9Xp1oXEYBU1jJ3bQ5k:b6zrrkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L744)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Do we want to specify 1 as the size of the channel?
</blockquote></details>

Good catch.
___
*[pkg/sql/colexec/routers_test.go, line 704 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PTivsB23OK6WLEDtj:-M9XpEBXAC8Dxak55ITY:b3b6v5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L704)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Should we also check that the last output in particular returned the metadata?
</blockquote></details>

It's not always going to be the last output in the slice because they all run concurrently.
___
*[pkg/sql/colexec/routers_test.go, line 759 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PTwsn0MdB3r6gu0pp:-M9Y4Pde8XxWKd8-XM4V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L759)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Is this still used?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/routers_test.go, line 984 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PdlFiDGec7XgVs7XY:-M9XpL1u3W1so_7ux6T0:b-namkuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L984)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Seems like a leftover artifact.
</blockquote></details>

Correct, any fmt.Println leftover is a a leftover artifact. Should've checked for this.
___
*[pkg/sql/colexec/routers_test.go, line 1060 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PeWrm9dorv8aKLApO:-M9XpgAH5aJw0X5i5dXD:bq50pc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L1060)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Probably we could use `isTerminationScenario` with 1.0 as probability.
</blockquote></details>

It requires us to create a new `rng` since this runs in a separate goroutine, so thought this would be less verbose.
___
*[pkg/sql/colexec/routers_test.go, line 1237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PfcqyBiZrowat8QCs:-M9Xpsg26rfeXuj2zi56:bijtwc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L1237)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: we probably can omit `_ = ` part, right?
</blockquote></details>

I like that this says that it's explicitly ignoring the return value but that's personal preference
___
*[pkg/sql/colexec/serial_unordered_synchronizer.go, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PnKDL0U8E09IiH3ll:-M9Y2X3K4u28il4X_f7l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e963bfed266c1025b02b5ed9c805fb2d16e0e36/pkg/sql/colexec/serial_unordered_synchronizer.go#L82)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto for the length check.
</blockquote></details>

Done.
___
*[pkg/sql/execinfrapb/data.go, line 395 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9Phaee7zZrm8F28L0j:-M9Xq7sk5UZuQTb6SmCg:b-4tijfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/execinfrapb/data.go#L395)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: we don't need to check the length, can just append unconditionally (I've just confirmed it here https://play.golang.org/p/d41I1wL7kIg)
</blockquote></details>

Cool! Thanks for checking.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> I think I'm going to tackle those in a separate PR though to keep the amount of code manageable.

SGTM.

> Reviewed 5 of 13 files at r3, 10 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-:-M9ZZPTpEYo0E3MsAELp:bcwglu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/routers.go, line 455 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PRBK01iJX3nIWhN7W:-M9ZzkBc0lzRWkDFuhTT:bhph56u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers.go#L455)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I started off with `hashRouterState` but I found it confusing because these states are explicitly about draining. For example, a hash router can have drained and exited the `Run` method but still be in `hashRouterStateRunning` until a somebody calls `drainMeta`. I'm happy to change it back, but wondering what you think.
</blockquote></details>

Hm, it's a good point, let's keep it this way. Also, I think it's worth adding your comment as the comment in the code, to the struct itself.
___
*[pkg/sql/colexec/routers.go, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9_-JMJC7w4Giga_Tcn:-M9_-JMJC7w4Giga_Tco:b-33xwkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8d911f414aaed882e16f404af4a3450f8cd8b1d/pkg/sql/colexec/routers.go#L39)):*
> ```Go
> 	execinfra.OpNode
> 	// initWithHashRouter passes a reference to the HashRouter that will be
> 	// pushing batches to this input.
> ```

nit: "this input" seems a little bit wrong because we have "routerOutput".
___
*[pkg/sql/colexec/routers.go, line 707 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9_2EqL6y7zsE6NSTQz:-M9_2EqL6y7zsE6NSTR-:b-eolgwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/65535e83fa4d30c401274776a8d22af4e5211edd/pkg/sql/colexec/routers.go#L707)):*
> ```Go
> 	r.bufferedMeta = append(r.bufferedMeta, r.metadataSources.DrainMeta(ctx)...)
> 	r.waitForMetadata <- r.bufferedMeta
> 	close(r.waitForMetadata)
> ```

Shouldn't the last router output be the one closing this channel once it has received the metadata?
___
*[pkg/sql/colexec/routers_test.go, line 704 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PTivsB23OK6WLEDtj:-M9_14di3Txw4j5fq1CI:b42jn76) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L704)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It's not always going to be the last output in the slice because they all run concurrently.
</blockquote></details>

Oh yeah, nvm.
___
*[pkg/sql/colexec/routers_test.go, line 1060 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PeWrm9dorv8aKLApO:-M9_1cLDDnen6MPNuG14:ba85ibv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L1060)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It requires us to create a new `rng` since this runs in a separate goroutine, so thought this would be less verbose.
</blockquote></details>

I see, nvm then.
___
*[pkg/sql/colexec/routers_test.go, line 1237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9PfcqyBiZrowat8QCs:-M9_1fkV4AMh5ND4WJK1:bl9l4q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c6d7d99e1e524ee73c1ba71a855460e48ff8b59/pkg/sql/colexec/routers_test.go#L1237)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I like that this says that it's explicitly ignoring the return value but that's personal preference
</blockquote></details>

Ok, sgtm.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50009)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/routers.go, line 707 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-M9_2EqL6y7zsE6NSTQz:-M9rEXZo-nIKw_NduMyI:bxfgqa6) ([raw file](https://github.com/cockroachdb/cockroach/blob/65535e83fa4d30c401274776a8d22af4e5211edd/pkg/sql/colexec/routers.go#L707)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Shouldn't the last router output be the one closing this channel once it has received the metadata?
</blockquote></details>

No, only senders should close channels as a signal that no more data will be sent on this channel. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
The build seems to be rather unhappy, but the code <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 12 of 12 files at r5, 10 of 10 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50009#-:-M9spPTyDs9pB2uBnnc6:b-umq6yn)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

asubiotto commented:
One failure was forgetting to finalize a refactor and the other was a legitimate failure when no metadata is drained from the parallel unordered synchronizer. Fixed both, let's see how the build is.

Commit:
colexec: make ParallelUnorderedSynchronizer implement DrainMeta

Previously, any components that would call DrainMeta downstream of the
ParallelUnorderedSynchronizer would call into an upstream goroutine. This
commit makes it so that any component downstream of the
ParallelUnorderedSynchronizer that requires metadata from one of the input
goroutines can now call DrainMeta on the ParallelUnorderedSynchronizer, who
will take care of returning DrainMeta accumulated from any input goroutines.

This commit also simplifies the other synchronizers so multiple metadata
sources can be collapsed into one. This is not necessary for DrainMeta and Next
to be called from a single goroutine but it does simplify the planning model.

Release note: None (no user-visible changes)

Commit:
colexec: move HashRouter DrainMeta call to router outputs

This commit implements a synchronized way in which metadata from the
HashRouter's run goroutine is propagated to the last router output to
DrainMeta.

Additionally, the case in which a router output encounters an error on Next
is reworked. The router output notifies the HashRouter to cancel all outputs
and returns the error directly to the Next goroutine.

Release note: None (no user-visible changes)

asubiotto commented:
bors r=yuzefovich

yuzefovich reviewed (approved):
> Reviewed 12 of 12 files at r7, 10 of 10 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50009)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->
